src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -0,0 +1,241 @@
+      (when http-client-cipher-suites
+        (.put puppet-server-config "cipher_suites" (into-array String http-client-cipher-suites)))
+      (.put puppet-server-config "profiler" profiler)
+      (.put puppet-server-config "environment_registry" env-registry)
+
+      (let [instance (jruby-schemas/map->JRubyPuppetInstance
+                       {:pool                 pool
+                        :id                   id
+                        :state                (atom {:request-count 0})
+                        :jruby-puppet         (.callMethod scripting-container
+                                                           ruby-puppet-class
+                                                           "new"
+                                                           (into-array Object
+                                                                       [puppet-config puppet-server-config])
+                                                           JRubyPuppet)
+                        :scripting-container  scripting-container
+                        :environment-registry env-registry})]
+        (.putLast pool instance)
+        instance))))
+
+(schema/defn ^:always-validate
+  get-pool-state :- jruby-schemas/PoolState
+  "Gets the PoolState from the pool context."
+  [context :- jruby-schemas/PoolContext]
+  @(:pool-state context))
+
+(schema/defn ^:always-validate
+  get-pool :- jruby-schemas/pool-queue-type
+  "Gets the JRubyPuppet pool object from the pool context."
+  [context :- jruby-schemas/PoolContext]
+  (:pool (get-pool-state context)))
+
+(schema/defn ^:always-validate
+  get-pool-size :- schema/Int
+  "Gets the size of the JRubyPuppet pool from the pool context."
+  [context :- jruby-schemas/PoolContext]
+  (get-in context [:config :max-active-instances]))
+
+(schema/defn borrow-without-timeout-fn :- jruby-schemas/JRubyPuppetBorrowResult
+  [pool :- jruby-schemas/pool-queue-type]
+  (.takeFirst pool))
+
+(schema/defn borrow-with-timeout-fn :- jruby-schemas/JRubyPuppetBorrowResult
+  [timeout :- schema/Int
+   pool :- jruby-schemas/pool-queue-type]
+  (.pollFirst pool timeout TimeUnit/MILLISECONDS))
+
+(schema/defn borrowed-poison-pill
+  [pool :- jruby-schemas/pool-queue-type
+   instance :- PoisonPill]
+  (.putFirst pool instance)
+  (throw (IllegalStateException.
+           "Unable to borrow JRuby instance from pool"
+           (:err instance))))
+
+(declare borrow-from-pool!*)
+
+(schema/defn borrowed-jruby-instance :- jruby-schemas/JRubyPuppetBorrowResult
+  [borrow-fn :- (schema/pred ifn?)
+   pool :- jruby-schemas/pool-queue-type
+   pool-context :- jruby-schemas/PoolContext
+   flush-instance-fn :- (schema/pred ifn?)
+   instance :- JRubyPuppetInstance]
+  (let [new-state (swap! (:state instance)
+                         update-in [:request-count] inc)
+        max-requests (get-in pool-context [:config :max-requests-per-instance])]
+    (if (and (pos? max-requests)
```

Comment:
Also, would that help with the full `flush-pool!` implementation?  If the `borrow-from-pool!*` didn't need to possibly flush an instance that was just borrowed, it would seem less likely we would end up possibly wasting one or more cases of creating and immediately flushing a new instance from `flush-pool!`.